Core ASLR

jenswi-linaro opened the pull request:
This PR introduces ASLR in OP-TEE core. ASLR with paging is currently not supported.

jforissier commented on core/arch/arm/arm.mk:
```diff
@@ -5,6 +5,10 @@ CFG_CORE_TZSRAM_EMUL_SIZE ?= 458752
 CFG_MMAP_REGIONS ?= 13
+ifeq ($(CFG_CORE_ASLR),y)
+# 1 extra region for the identity map
+CFG_MMAP_REGIONS := $(CFG_MMAP_REGIONS) + 1
```
This syntax is problematic because I think the two following commands will behave differently: `CFG_MMAP_REGIONS=13 make` and `make CFG_MMAP_REGIONS=13` (and until now we have avoided that).
Also it is a bit weird that the end value can be different than the one explicitly provided.

How about something like this instead?
```make
ifeq ($(CFG_CORE_ASLR),y)
IDMAP_REGIONS := 1
else
IDMAP_REGIONS := 0
endif
CFG_MMAP_REGIONS ?= 13 + $(IDMAP_REGIONS)
```

jenswi-linaro replied:
Well spotted. Perhaps this should be handled directly with an ifdef where `static_memory_map` is declared instead to keep things simple.

jenswi-linaro commented on core/arch/arm/kernel/generic_boot.c:
```diff
@@ -1202,3 +1202,31 @@ struct ns_entry_context *generic_boot_core_hpen(void)
+		DMSG("Cannot find /secure-chosen");
+		return 0;
+	}
+	seed = fdt_getprop(fdt, offs, "kaslr-seed", &len);
```
Agree, I'll look into posting something.
The intent is to have the equivalent of the non-secure chosen/kaslr-seed.

jenswi-linaro commented on mk/config.mk:
Agree

jenswi-linaro commented:
I'll print the base address in the stack dump code.

jbech-linaro commented:
> I'll print the base address in the stack dump code.

This must be under some kind of debug flag I suppose? I.e., even though you're not getting anywhere after a panic (and next run you have a new address offset), it's better if we don't expose any addresses at all on non-debug builds. Agree?

jenswi-linaro commented:
Isn't any secure print per definition is a debug print?

jbech-linaro commented:
> Isn't any secure print per definition is a debug print?

Yes, no, depends :) If you're printing it as a `DMSG`, then we're on the safe side. EMSG is something that I find it likely that vendors would like to have enabled on deployed devices.

jenswi-linaro commented:
For debug purposes it doesn't make sense to do a stack trace without intentionally revealing the base address.

If stack traces should be done at error level or some other level could be discussed, but I think we could take that in another PR.

jenswi-linaro commented:
Update

jenswi-linaro commented:
By the way, it seems that we don't set `__HAVE_BUILTIN_OVERFLOW` for Clang.

jforissier commented:
> By the way, it seems that we don't set `__HAVE_BUILTIN_OVERFLOW` for Clang

We definitely should ;-)

jbech-linaro commented:
@jforissier I'm hijacking the thread regarding clang. What I can see we have no automatic testing of clang, right? If that is true, it'd be good to at least start with a single build that uses clang. This is something that we could expand later on. We have some tiny wiggle room left on either Travis or Shippable?

jforissier commented:
@jbech-linaro 

> What I can see we have no automatic testing of clang, right? 

Actually we do ;-) https://github.com/OP-TEE/optee_os/commit/f417c69559285e1ce5060497139dd7bc6c7d5faa

jbech-linaro commented:
> Actually we do ;-) [f417c69](https://github.com/OP-TEE/optee_os/commit/f417c69559285e1ce5060497139dd7bc6c7d5faa)

Huga! I grepped for clang and couldn't see it and now I realize why, nowadays I use "[ripgrep](https://blog.burntsushi.net/ripgrep/)" instead of standard grep and it doesn't include all files by default. Well, all good then, thanks and sorry for the noise!

jforissier commented:
@jenswi-linaro 

> I'll print the base address in the stack dump code.

Good! The corresponding patch for `symbolize.py` is here: https://github.com/jforissier/optee_os/commit/83c5b8362ac0. Feel free to pick it up and push it here.

jenswi-linaro commented:
Thanks @jforissier, I've pushed the patch.

jforissier commented:
@jenswi-linaro 

> By the way, it seems that we don't set `__HAVE_BUILTIN_OVERFLOW` for Clang.

For the record, it is not as simple as it seems. Enabling those builtins with Clang results in error:
```
undefined reference to `__muloti4'
```
According to this [comment](https://github.com/android/ndk/issues/295#issuecomment-277098658) it should be fixed by linking against the compiler-rt library. Unfortunately the method to cross-compile `compiler-rt` is [quite complicated](https://llvm.org/docs/HowToCrossCompileBuiltinsOnArm.html), so I decided to stop my investigations at this point. (No, it is not enough to add `compiler-rt` to `LLVM_ENABLE_PROJECTS`).

jenswi-linaro commented:
@jforissier, thanks for the heads up. Good thing we have the fallback option then. :-)

jforissier commented on core/arch/arm/kernel/kern.ld.S:
```diff
@@ -172,6 +172,12 @@ SECTIONS
+	 * .got is normally read/write but once relocation is done it's
+	 * more safe to make it read-only.
+	 */
+	__rodata_end = .;
```
So now `__rodata_start` / `__rodata_end` do not represent the start and end of the `.rodata` section anymore, but the start and end of the read-only sections (well, those we want to be read-only after initializations). It's a bit misleading. How about `__ro_sections_start` / `__ro_sections_end` instead?

jenswi-linaro replied:
I'll fix.

jenswi-linaro commented on core/arch/arm/kernel/generic_boot.c:
```diff
@@ -1202,3 +1202,31 @@ struct ns_entry_context *generic_boot_core_hpen(void)
+	}
+	seed = fdt_getprop(fdt, offs, "kaslr-seed", &len);
+	if (!seed || len != sizeof(*seed)) {
+		DMSG("Cannot find kaslr-seed");
```
I'll fix

jenswi-linaro commented on core/arch/arm/kernel/generic_boot.c:
```diff
@@ -1202,3 +1202,31 @@ struct ns_entry_context *generic_boot_core_hpen(void)
 }
 #endif
+
+#ifdef CFG_CORE_ASLR
```
I'll fix.

jenswi-linaro commented on core/arch/arm/kernel/generic_entry_a32.S:
```diff
@@ -512,10 +531,48 @@ UNWIND(	.fnstart)
+1:	ldmia	r1!, {r4-r5}		/* r4 = r_offset, r5 = r_info */
+	/* We're only expecting R_ARM_RELATIVE relocations. */
+	cmp	r5, #R_ARM_RELATIVE
+	bne	.
```
We can add a post link time check. In fact I'm currently writing a new script which will replace objdump and scripts/gen_hashed_bin.py. I needed some more flexibility in where the relocation information it's stored in the pager case. However, that should probably go into a separate PR.

jforissier replied:
Sounds good.

jenswi-linaro commented on core/arch/arm/kernel/generic_entry_a64.S:
```diff
@@ -137,11 +138,13 @@ clear_nex_bss:
+
+	bl	__get_core_pos
+	bl	enable_mmu
+	/* Virtual addresses may have changed, reinitialize console */
```
No, this is old stuff needed without ASLR.

jforissier replied:
I don't understand why it is needed, with or without ASLR actually. The console code uses a PA when the MMU is disabled, then resolves the PA to a VA (and caches it) when the MMU is enabled (thanks to `io_pa_or_va()`). So, as long as the VA does not change, we should be good. 

What do you mean by "Virtual addresses may have changed"? What could change the mapping once it has been established and the MMU turned on?

I must be missing something because if I remove that `console_init`, all is fine w/o ASLR but the initialization hangs with ASLR :-/

jenswi-linaro replied:
The problem is with `register_serial_console()` which has previously registered a physical address, now that virtual addresses are different from physical ones we need to register again.

I'll guard the call with `CFG_CORE_ASLR`.

jforissier replied:
That's it... I wasn't looking at the right place, I was too focused on this `io_pa_or_va()` thing :-/ Would you mind updating the comment to something like this?
```C 
 /*
  * Reinitialize console, since register_serial_console() has previously
  * registered a PA and with ASLR the VA is different from the PA.
  */
```

jenswi-linaro replied:
OK

jenswi-linaro commented on core/arch/arm/kernel/generic_entry_a32.S:
```diff
@@ -459,13 +440,13 @@ shadow_stack_access_ok:
+	bl	__get_core_pos
+	bl	enable_mmu
+	/* Virtual addresses may have changed, reinitialize console */
+	bl	console_init
```
No, this is old stuff needed without ASLR.

jenswi-linaro commented:
Addressing comments.
With all the review patches it's starting be a bit messy. OK if I squash in the fixes and apply tags where applicable?

jforissier commented on core/arch/arm/kernel/kern.ld.S:
```diff
@@ -85,6 +85,10 @@ SECTIONS
 		KEEP(*(.text._start))
+		__identity_map_init_start = .;
+		*(.identity_map .identity_map.* \
+		  .text.__get_core_pos .text.get_core_pos_mpidr)
```
> that's weak and we don't want to add that burden on the platforms which overrides `get_core_pos_mpidr()`

Ah, I see. Fair enough. Then I wouldn't mind having a comment here ;-)

As for `__get_core_pos()` I have a slight preference for not making it a special case since there is no real reason, i.e., use `FUNC func , : , .identity_map`. But not a strong opinion.

jenswi-linaro replied:
I'll fix

jforissier commented:
> OK if I squash in the fixes and apply tags where applicable?

Fine by me.

jforissier commented on core/arch/arm/include/mm/core_mmu.h:
```diff
@@ -141,6 +143,7 @@ static inline const char *teecore_memtype_name(enum teecore_memtypes type)
 		[MEM_AREA_TEE_RAM_RW] = "TEE_RAM_RW",
 		[MEM_AREA_NEX_RAM_RW] = "NEX_RAM_RW",
 		[MEM_AREA_TEE_ASAN] = "TEE_ASAN",
+		[MEM_AREA_IDENTITY_MAP_RX] = "MEM_AREA_IDENTITY_MAP_RX",
```
Sorry I missed that in my review... should be `= "IDENTITY_MAP_RX"`

jenswi-linaro replied:
I'll fix

jenswi-linaro commented:
I'll drop "core: pager: avoid *_size globals from linker script", to be introduced with support for paging instead.

jenswi-linaro commented:
Squashed review commits and applied tags.

jenswi-linaro commented:
It seems that pycodestyle is unhappy with the changes to `scripts/symbolize.py`, see https://travis-ci.org/jenswi-linaro/optee_os/builds/600720717

jforissier commented:
> It seems that pycodestyle is unhappy [...]

Fixed in new tip of my branch core-aslr-symbolize, here: https://github.com/jforissier/optee_os/commit/1e8659ae.

jenswi-linaro commented:
Rebased on master and replace "symbolize.py: add support for TEE core ASLR" with jforissier@1e8659a.

jenswi-linaro commented:
Added a fix for "core: refactor core_init_mmu_map() and helpers"

jforissier commented on core/arch/arm/include/kernel/generic_boot.h:
```diff
@@ -53,4 +54,6 @@ void *get_embedded_dt(void);
 /* Returns external DTB if present, otherwise NULL */
 void *get_external_dt(void);
 
+unsigned long get_aslr_seed(void *fdt);
```
We need a stub for `!CFG_DT` otherwise there is a link error (HiKey960 for instance). Of course it doesn't make much sense to enable ASLR without DT since the only source for the seed is the DT, but it's still useful for testing and also we could have another (platform-specific) seed source.

jforissier replied:
...actually we need it in `generic_boot.c`, not here, since it's called from assembler.

jenswi-linaro replied:
Good point, I'll fix

jforissier commented:
Tested successfully on HiKey960 (at commit 0ef4ce3 plus a small patch to provide a fixed seed via `get_aslr_seed()` since HiKey has `CFG_DT=n`). So:

`Tested-by: Jerome Forissier <jerome@forissier.org> (HiKey960)`

jenswi-linaro commented:
Did a force update to update the commit message for the get_aslr_seed() update

jenswi-linaro commented:
Another force update to add a S-o-B to "Fix warning in fallback SUB_OVERFLOW() macro "

jenswi-linaro commented:
The following commits are missing acked-by or reviewed-by tags:
"core: refactor core_init_mmu_map() and helpers"
"core: add get_aslr_seed()"
"core: arm: increase CFG_MMAP_REGIONS with ASLR"
"core: introduce CFG_CORE_ASLR"
"Fix warning in fallback SUB_OVERFLOW() macro"
"symbolize.py: add support for TEE core ASLR"

jenswi-linaro commented on mk/config.mk:
We can do that when it's compatible with paging also. It's not that far off, two weeks until the pull request perhaps (I'm on vacation next week).

jenswi-linaro commented on core/arch/arm/kernel/generic_boot.c:
```diff
@@ -1233,4 +1234,10 @@ unsigned long get_aslr_seed(void *fdt)
+#else /*!CFG_DT*/
+unsigned long __weak get_aslr_seed(void *fdt __unused)
+{
+	return 0;
```
Makes sense, I'll fix.

jenswi-linaro commented on core/arch/arm/arm.mk:
```diff
@@ -138,6 +134,11 @@ core-platform-cflags += $(platform-cflags-debug-info)
 
+ifeq ($(CFG_CORE_ASLR),y)
+core-platform-cflags += -fpie -fvisibility=hidden
+core-platform-aflags += -fvisibility=hidden
```
Doesn't seem to be needed, I'll remove it.

jforissier replied:
Ha! Even better then.

jenswi-linaro commented:
Fixed up commits and added tags.
Dropped "core: arm: increase CFG_MMAP_REGIONS with ASLR"
Added ASLR to some shippable builds
Added "[review] core: add support for CFG_CORE_ASLR" to hopefully take care of the clang linker problem.

jforissier commented on mk/clang.mk:
```diff
@@ -19,6 +19,8 @@ ifeq ($(sm)-$(CFG_WITH_PAGER),core-y)
 # Anyway, use GNU.ld for the moment.
 LDcore		:= $(CROSS_COMPILE_$(sm))ld
+else
+comp-ldflags-allow-text-reloc-clang := -Wl,-z,notext
```
Should be `-z notext` because it's linker flags, not compiler flags for the linker.

With that fixed, I get:
```
ld.lld: error: section: .ctors is not contiguous with other relro sections
ld.lld: error: section: .dtors is not contiguous with other relro sections
```

...which can be fixed by moving `.ctors` and `.dtors` just after `.plt`. Then the build goes fine but the board hangs on boot. I'll take a closer look.

jforissier replied:
FYI, the Clang build works OK on my HiKey960 if GNU ld is used.

jenswi-linaro replied:
QEMU virt v7 works fine with clang with the ctors/dtors fix applied.
QEMU virt v8 crashes due to stack overrun (filling translation tables with junk) on the TMP stack during initialization.

It seems it's `kprintf()` and friends whos using a lot of stack.

jforissier replied:
Same situation with on HiKey960: 32-bit OK, 64-bit hangs during initialization (tested commit 235cf33c with clang).

jforissier commented:
For "core: kern.ld.S: make ctors/dtors contiguous with other relro sections":
`Reviewed-by: Jerome Forissier <jerome@forissier.org>`

(minor: second line of error message: s/section./sections/)

jenswi-linaro commented:
tag applied and "core: kern.ld.S: make ctors/dtors contiguous with other relro sections" moved just after the other kern.ld.S commits. "core: add support for CFG_CORE_ASLR" fixed up with the "[review]" commits.

jenswi-linaro commented:
The following commits are missing acked-by or reviewed-by tags:
- "ci: shippable: build with CFG_CORE_ASLR=y"
- "core: add support for CFG_CORE_ASLR"
- "core: refactor core_init_mmu_map() and helpers"

jforissier commented:
"ci: shippable: build with CFG_CORE_ASLR=y" is fine by me:
`Acked-by: Jerome Forissier <jerome@forissier.org>`

As for the other two commits, TBH they are a bit too much outside my comfort zone to give a meaningful feedback :-/

jenswi-linaro commented:
Thanks @jforissier, I'm hoping that someone else (hello @etienne-lms :-) ) can review the two remaining commits.

etienne-lms commented:
yes, i'll do. Sorry for my poor participation to the forum these days...

jenswi-linaro commented:
Thanks

etienne-lms commented on core/arch/arm/mm/core_mmu.c:
```diff
@@ -1054,12 +1076,27 @@ void core_init_mmu_map(void)
+ *
+ * this routine sets the static default tee core mapping.
+ *
+ * If an error happened: core_init_mmu_map is expected to reset.
```
s/reset/panic/

etienne-lms commented on core/arch/arm/mm/core_mmu.c:
```diff
@@ -1054,12 +1076,27 @@ void core_init_mmu_map(void)
+/*
+ * core_init_mmu_map - init tee core default memory mapping
+ *
+ * this routine sets the static default tee core mapping.
```
s/this/This/
s/tee/TEE/g

etienne-lms commented on core/arch/arm/mm/core_mmu.c:
```diff
@@ -893,153 +896,172 @@ static void init_mem_map(struct tee_mmap_region *memory_map, size_t num_elems)
+{
+	struct tee_mmap_region *m = NULL;
+
+	for (m = map; core_mmap_is_end_of_table(m); m++) {
```
```patch
-	for (m = map; core_mmap_is_end_of_table(m); m++) {
+	for (m = map; !core_mmap_is_end_of_table(m); m++) {

etienne-lms commented on core/arch/arm/mm/core_mmu.c:
```diff
@@ -893,153 +896,172 @@ static void init_mem_map(struct tee_mmap_region *memory_map, size_t num_elems)
+	const struct tee_mmap_region *mm_b = b;
+	int rc = 0;
+
+	if (!rc)
```
minor: could remove this test

jenswi-linaro commented:
Force push to fix up "core: refactor core_init_mmu_map() and helpers" according to the comments and apply the tag.

etienne-lms commented on .shippable.yml:
```diff
@@ -37,7 +37,7 @@ build:
     - _make CFG_WITH_PAGER=y CFG_CRYPTOLIB_NAME=mbedtls CFG_CRYPTOLIB_DIR=lib/libmbedtls
     - _make CFG_WITH_PAGER=y CFG_WITH_LPAE=y
-    - _make CFG_WITH_LPAE=y
+    - _make CFG_WITH_LPAE=y CFG_CORE_ASLR=y
```
Would be nice to have a build in 32bit MMU with ASLR, i.e here `_make CFG_CORE_ASLR=y`.

jenswi-linaro commented:
Force push to fixup "ci: shippable: build with CFG_CORE_ASLR=y" according to comments and applied the tag.

jenswi-linaro commented:
Force push to fixup "core: add support for CFG_CORE_ASLR" according to comments and applied the tag.

etienne-lms commented:
Generic implementation question: this adds an offset computed from the seed read from the DT and some core static data (va width, virtual address space size). So it seems to me that core is always loaded at the same place assuming seed value from DT is the same. Did I miss something?

jforissier commented:
@etienne-lms seed is supposed to be random, filled by the previous stage bootloader.

jforissier commented:
@etienne-lms thanks for your help :+1: now this can be merged.

jenswi-linaro commented:
Thanks @jforissier and @etienne-lms :-)

jbech-linaro commented:
Good work @jenswi-linaro @jforissier @etienne-lms , this is yet another an important component that raises the "security bar" yet another notch.
